src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py

```diff
@@ -0,0 +1,1120 @@
+
+    HunyuanDiT uses two text encoders: [mT5](https://huggingface.co/google/mt5-base) and [bilingual CLIP](fine-tuned by ourselves)
+    
+    Args:
+        vae ([`AutoencoderKL`]):
+            Variational Auto-Encoder (VAE) Model to encode and decode images to and from latent representations.
+            We use `sdxl-vae-fp16-fix`.    
+        text_encoder (Optional[`~transformers.BertModel`, `~transformers.CLIPTextModel`]):
+            Frozen text-encoder ([clip-vit-large-patch14](https://huggingface.co/openai/clip-vit-large-patch14)).
+            HunyuanDiT uses a fine-tuned [bilingual CLIP].
+        tokenizer (Optional[`~transformers.BertTokenizer`, `~transformers.CLIPTokenizer`]):
+            A `BertTokenizer` or `CLIPTokenizer` to tokenize text.
+        transformer ([`HunyuanDiT2DModel`]):
+            The HunyuanDiT model designed by Tencent Hunyuan.
+        embedder_t5 (`MT5Embedder`):
+            The mT5 embedder. Specifically, it is 't5-v1_1-xxl'.
+        scheduler ([`SchedulerMixin`]):
+            A scheduler to be used in combination with HunyuanDiT to denoise the encoded image latents.
+    """
+
+    model_cpu_offload_seq = "text_encoder->embedder_t5->tokenizer->tokenizer_t5->transformer->vae"
+    _optional_components = ["safety_checker", "feature_extractor", "embedder_t5", "tokenizer_t5"]
+    _exclude_from_cpu_offload = ["safety_checker"]
+
+    def __init__(
+            self,
+            vae: AutoencoderKL,
+            text_encoder: BertModel,
+            tokenizer: BertTokenizer,
+            transformer: HunyuanDiT2DModel,
+            scheduler: DDPMScheduler,
+            safety_checker: StableDiffusionSafetyChecker,
+            feature_extractor: CLIPImageProcessor,
+            requires_safety_checker: bool = True,
+            progress_bar_config: Dict[str, Any] = None,
+            embedder_t5=T5EncoderModel,
+            tokenizer_t5=MT5Tokenizer,
+            infer_mode='torch',
+    ):
+        super().__init__()
+
+        # ========================================================
+        self.infer_mode = infer_mode
+
+        # ========================================================
+        if progress_bar_config is None:
+            progress_bar_config = {}
+        if not hasattr(self, '_progress_bar_config'):
+            self._progress_bar_config = {}
+        self._progress_bar_config.update(progress_bar_config)
+        # ========================================================
+        
+        self.register_modules(
+            vae=vae,
+            text_encoder=text_encoder,
+            tokenizer=tokenizer,
+            tokenizer_t5=tokenizer_t5,
+            transformer=transformer,
+            scheduler=scheduler,
+            safety_checker=safety_checker,
+            feature_extractor=feature_extractor,
+            embedder_t5=embedder_t5,
+        )
+
+        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
```

Comment:
Can you elaborate what issue you're facing here? This is anti-pattern in the library.

Reply:
There is an extra 'pooler' module initialized in BertModel. Don't know how to configure correctly to remove it and fit in `from_pretrained`. In our pretrained Bert, we turned this layer off at intialization.

Reply:
ohh I see https://github.com/huggingface/transformers/blob/965e98dc54c4853b34b3a49810f24856a9fcdba8/src/transformers/models/bert/modeling_bert.py#L976

@sayakpaul can you look into how we can make it work with `from_pretrained()`?
do we need to make a wrapper of bert model inside diffusers?

Reply:
Could we do something like this when passing the `BertModel`?

```python
from transformers import BertModel

bert_ckpt = "..."
bert_model = BertModel.from_pretrained(bert_ckpt, add_pooling_layer=False)
```

This way, we should be able to obtain the right variant of the `BertModel` that we want and serialize it along with the other components of the pipeline.

Reply:
Doesn't this work?

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline

ckpt_id = "XCLiu/HunyuanDiT-0523"
model = BertModel.from_pretrained(ckpt_id, add_pooling_layer=False, subfolder="text_encoder")

pipeline = HunyuanDiTPipeline.from_pretrained(
    ckpt_id, text_encoder=model, ...
)
```

@gnobitab could you give this a try without the manual `to_empty()` operation?

Reply:
Just a disclaiming that when you load `BertModel` like the above, you would notice:

```bash
Some weights of the model checkpoint at XCLiu/HunyuanDiT-0523 were not used when initializing BertModel: ['pooler.dense.bias', 'pooler.dense.weight']
```

You mentioned the `pooler` layers shouldn't matter. So, LMK.

Reply:
we should have this work 
```python
pipeline = HunyuanDiTPipeline.from_pretrained(ckpt_id)
```

Reply:
Yes, this would work after the pipeline is pushed with the modifications I suggested.

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline

ckpt_id = "XCLiu/HunyuanDiT-0523"
model = BertModel.from_pretrained(ckpt_id, add_pooling_layer=False, subfolder="text_encoder")

pipeline = HunyuanDiTPipeline.from_pretrained(
    ckpt_id, text_encoder=model, ...
)
pipeline.push_to_hub(...)
```

And then  

```python
pipeline = HunyuanDiTPipeline.from_pretrained(ckpt_id)
```

should work without `to_empty()`.

Reply:
ohhh ok!

Reply:
Thanks for the suggestion! This code snippet works without `to_empty`. 
But after I saved the pipeline to with 
```py
from transformers import BertModel
bert_model = BertModel.from_pretrained("XCLiu/HunyuanDiT-0523", add_pooling_layer=False, subfolder="text_encoder")

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model, torch_dtype=torch.float32)
pipe.to('cuda')
pipe.save_pretrained("../HunyuanDiT-ckpt")
``` 
and reload it with
```py
pipeline = HunyuanDiTPipeline.from_pretrained("../HunyuanDiT-ckpt")
pipe.to('cuda')
```
The same error appears again...
<img width="1347" alt="image" src="https://github.com/huggingface/diffusers/assets/1157982/f69c2041-a25f-4dd6-b0df-a63681cbefcd">

Reply:
I am able to successfully do:

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline, HunyuanDiT2DModel
from huggingface_hub import hf_hub_download
import safetensors
import torch

def load_transformer():
    model_config = HunyuanDiT2DModel.load_config("XCLiu/HunyuanDiT-0523", subfolder="transformer")
    model = HunyuanDiT2DModel.from_config(model_config)

    ckpt_path = hf_hub_download(
        "XCLiu/HunyuanDiT-0523",
        filename ="diffusion_pytorch_model.safetensors",
        subfolder="transformer",
    )
    state_dict = safetensors.torch.load_file(ckpt_path)

    num_layers = 40
    for i in range(num_layers):
        
        # attn1
        # Wkqv -> to_q, to_k, to_v
        q, k, v = torch.chunk(state_dict[f"blocks.{i}.attn1.Wqkv.weight"], 3, dim=0)
        q_bias, k_bias, v_bias = torch.chunk(state_dict[f"blocks.{i}.attn1.Wqkv.bias"], 3, dim=0)
        state_dict[f"blocks.{i}.attn1.to_q.weight"] = q
        state_dict[f"blocks.{i}.attn1.to_q.bias"] = q_bias
        state_dict[f"blocks.{i}.attn1.to_k.weight"] = k
        state_dict[f"blocks.{i}.attn1.to_k.bias"] = k_bias
        state_dict[f"blocks.{i}.attn1.to_v.weight"] = v
        state_dict[f"blocks.{i}.attn1.to_v.bias"] = v_bias
        state_dict.pop(f"blocks.{i}.attn1.Wqkv.weight")
        state_dict.pop(f"blocks.{i}.attn1.Wqkv.bias")
        
        # q_norm, k_norm -> norm_q, norm_k
        state_dict[f"blocks.{i}.attn1.norm_q.weight"] = state_dict[f"blocks.{i}.attn1.q_norm.weight"]
        state_dict[f"blocks.{i}.attn1.norm_q.bias"] = state_dict[f"blocks.{i}.attn1.q_norm.bias"]
        state_dict[f"blocks.{i}.attn1.norm_k.weight"] = state_dict[f"blocks.{i}.attn1.k_norm.weight"]
        state_dict[f"blocks.{i}.attn1.norm_k.bias"] = state_dict[f"blocks.{i}.attn1.k_norm.bias"]

        state_dict.pop(f"blocks.{i}.attn1.q_norm.weight")
        state_dict.pop(f"blocks.{i}.attn1.q_norm.bias")
        state_dict.pop(f"blocks.{i}.attn1.k_norm.weight")
        state_dict.pop(f"blocks.{i}.attn1.k_norm.bias")

        # out_proj -> to_out
        state_dict[f"blocks.{i}.attn1.to_out.0.weight"] = state_dict[f"blocks.{i}.attn1.out_proj.weight"]
        state_dict[f"blocks.{i}.attn1.to_out.0.bias"] = state_dict[f"blocks.{i}.attn1.out_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn1.out_proj.weight")
        state_dict.pop(f"blocks.{i}.attn1.out_proj.bias")

        # attn2
        # kq_proj -> to_k, to_v
        k, v = torch.chunk(state_dict[f"blocks.{i}.attn2.kv_proj.weight"], 2, dim=0)
        k_bias, v_bias = torch.chunk(state_dict[f"blocks.{i}.attn2.kv_proj.bias"], 2, dim=0)
        state_dict[f"blocks.{i}.attn2.to_k.weight"] = k
        state_dict[f"blocks.{i}.attn2.to_k.bias"] = k_bias
        state_dict[f"blocks.{i}.attn2.to_v.weight"] = v
        state_dict[f"blocks.{i}.attn2.to_v.bias"] = v_bias
        state_dict.pop(f"blocks.{i}.attn2.kv_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.kv_proj.bias")
        
        # q_proj -> to_q
        state_dict[f"blocks.{i}.attn2.to_q.weight"] = state_dict[f"blocks.{i}.attn2.q_proj.weight"]
        state_dict[f"blocks.{i}.attn2.to_q.bias"] = state_dict[f"blocks.{i}.attn2.q_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn2.q_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.q_proj.bias")
        
        # q_norm, k_norm -> norm_q, norm_k
        state_dict[f"blocks.{i}.attn2.norm_q.weight"] = state_dict[f"blocks.{i}.attn2.q_norm.weight"]
        state_dict[f"blocks.{i}.attn2.norm_q.bias"] = state_dict[f"blocks.{i}.attn2.q_norm.bias"]
        state_dict[f"blocks.{i}.attn2.norm_k.weight"] = state_dict[f"blocks.{i}.attn2.k_norm.weight"]
        state_dict[f"blocks.{i}.attn2.norm_k.bias"] = state_dict[f"blocks.{i}.attn2.k_norm.bias"]

        state_dict.pop(f"blocks.{i}.attn2.q_norm.weight")
        state_dict.pop(f"blocks.{i}.attn2.q_norm.bias")
        state_dict.pop(f"blocks.{i}.attn2.k_norm.weight")
        state_dict.pop(f"blocks.{i}.attn2.k_norm.bias")

        # out_proj -> to_out
        state_dict[f"blocks.{i}.attn2.to_out.0.weight"] = state_dict[f"blocks.{i}.attn2.out_proj.weight"]
        state_dict[f"blocks.{i}.attn2.to_out.0.bias"] = state_dict[f"blocks.{i}.attn2.out_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn2.out_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.out_proj.bias")

        # switch norm 2 and norm 3
        norm2_weight = state_dict[f"blocks.{i}.norm2.weight"]
        norm2_bias = state_dict[f"blocks.{i}.norm2.bias"]
        state_dict[f"blocks.{i}.norm2.weight"] = state_dict[f"blocks.{i}.norm3.weight"]
        state_dict[f"blocks.{i}.norm2.bias"] = state_dict[f"blocks.{i}.norm3.bias"]
        state_dict[f"blocks.{i}.norm3.weight"] = norm2_weight
        state_dict[f"blocks.{i}.norm3.bias"] = norm2_bias

    model.load_state_dict(state_dict)
    return model


bert_model = BertModel.from_pretrained(
    "XCLiu/HunyuanDiT-0523", add_pooling_layer=False, subfolder="text_encoder"
)
model = load_transformer()
pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model)
pipe.save_pretrained("HunyuanDiT-ckpt")

del pipe 

pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")
```

With the following change:

```diff
diff --git a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
index 381ded58..f373826e 100644
--- a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
+++ b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
@@ -408,7 +408,7 @@ class HunyuanDiTPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoa
             embedder_t5=embedder_t5,
         )
 
-        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
+        # self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
 
         if safety_checker is None and requires_safety_checker:
             logger.warning(

```

LMK if that works?

Reply:
It doesn't work for me ... I copy-paste your test code and it still shows this:

<img width="1363" alt="image" src="https://github.com/huggingface/diffusers/assets/1157982/59f9dfa3-cb09-4d0c-9eeb-708b9705d716">

So (1) Sayak could you share your package versions? (2) could you try `pipe.to('cuda')` after `pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")`. It's `pipe.to('cuda')` that pops errors. Thanks.

Reply:
I think you didn't apply the patch I mentioned in my comment. Here it is again for your reference: 

```diff
diff --git a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
index 381ded58..f373826e 100644
--- a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
+++ b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
@@ -408,7 +408,7 @@ class HunyuanDiTPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoa
             embedder_t5=embedder_t5,
         )
 
-        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
+        # self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
 
         if safety_checker is None and requires_safety_checker:
             logger.warning(
```

Reply:
it looks like this patch only make `self.text_encoder.pooler.to_empty(device='cpu')` a comment? if so, I already did that.

Reply:
That is quite weird then. 

Can you check if you are on the latest version of `transformers` and `accelerate`?

Reply:
I'm using `transformers==0.41.1` and `accelerate==0.30.1`. let me find out what's wrong. it is weird.

Reply:
I finally get it to work! With the following codes:

```py
from transformers import BertModel
bert_model = BertModel.from_pretrained("XCLiu/HunyuanDiT-0523", add_pooling_layer=True, subfolder="text_encoder")

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model, torch_dtype=torch.float32)
pipe.to('cuda')
pipe.save_pretrained("HunyuanDiT-ckpt")

del pipe

pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")
pipe.cuda()
```

In this way, the random-initialized weights in the 'pooler' layer with be stored, but it will not affect the results and kept the code snippet simple as `pipe = HunyuanDiTPipeline.from_pretrained('xxx')`. Thank you Sayak for the help!